File: backend/routers/leaderboard.js
'use strict';
const router = require('express').Router();
const db     = require('../core/database');
const { requireAuth } = require('../middleware/auth');

// GET /api/leaderboard/
router.get('/', async (req, res, next) => {
  try {
    const limit = Math.min(parseInt(req.query.limit, 10) || 50, 200);
    const entries = (await db.getLeaderboard()).slice(0, limit).map((row, i) => ({ ...row, rank: i + 1 }));
    res.json({ leaderboard: entries, count: entries.length });
  } catch (err) { next(err); }
});

// POST /api/leaderboard/submit
router.post('/submit', requireAuth, async (req, res, next) => {
  try {
    const { section, score, passed_tests, total_tests, time_taken = 0 } = req.body;
    if (![1,2,3,4].includes(section))
      return res.status(422).json({ detail: 'section must be 1–4' });
    if (score < 0 || score > 420)
      return res.status(422).json({ detail: 'score out of range' });

    await db.upsertScore({
      username: req.user.sub, section, score,
      passedTests: passed_tests, totalTests: total_tests, timeTaken: time_taken,
      fullName: req.user.full_name || '', college: req.user.college || '', team: req.user.team || '',
    });
    res.status(201).json({ message: 'Score recorded', section, score });
  } catch (err) { next(err); }
});

// GET /api/leaderboard/me
router.get('/me', requireAuth, async (req, res, next) => {
  try {
    const row = await db.getUserScores(req.user.sub);
    res.json({ username: req.user.sub, scores: row || {} });
  } catch (err) { next(err); }
});

module.exports = router;

File: backend/routers/leaderboard.py
"""
Leaderboard router — view rankings, submit scores.
"""

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from typing import Optional

from core.database import get_leaderboard, upsert_score, get_user_scores
from core.deps import get_current_user

router = APIRouter()


# ── Schemas ───────────────────────────────────────────────────────────────────

class ScoreSubmission(BaseModel):
    section: int           # 1–4
    score: float
    passed_tests: int
    total_tests: int
    time_taken: Optional[float] = 0.0


# ── Endpoints ─────────────────────────────────────────────────────────────────

@router.get("/")
def leaderboard(limit: int = 50):
    """Public leaderboard — top N participants."""
    entries = get_leaderboard()
    entries = entries[:min(limit, 200)]
    for i, row in enumerate(entries, 1):
        row["rank"] = i
    return {"leaderboard": entries, "count": len(entries)}


@router.post("/submit", status_code=status.HTTP_201_CREATED)
def submit_score(body: ScoreSubmission, user: dict = Depends(get_current_user)):
    """Submit a test-run score for the authenticated participant."""
    if body.section not in (1, 2, 3, 4):
        raise HTTPException(status_code=422, detail="section must be 1–4")
    if not (0 <= body.score <= 420):
        raise HTTPException(status_code=422, detail="score out of range")

    upsert_score(
        username=user["sub"],
        section=body.section,
        score=body.score,
        passed_tests=body.passed_tests,
        total_tests=body.total_tests,
        time_taken=body.time_taken,
        full_name=user.get("full_name", ""),
        college=user.get("college", ""),
        team=user.get("team", ""),
    )
    return {"message": "Score recorded", "section": body.section, "score": body.score}


@router.get("/me")
def my_scores(user: dict = Depends(get_current_user)):
    """Return current user's leaderboard row."""
    row = get_user_scores(user["sub"])
    return {"username": user["sub"], "scores": row or {}}
